### DISCUSSION TITLE
Can we use gossipsub and request-response protocols together ? It wasn't working fine for me

### DISCUSSION THREAD

**Discussion by vr16x** [libp2p/rust-libp2p#5141] [Q&A]
I am building an p2p system with mdns, gossipsub and request-response protocols. The request-response is not sending any request or response when I am publishing the request response data in gossipsub protocol. I need some assisstance over here.

I am using Rust Libp2p stack and the version is 53.2

**Comment by dariusc93** [comment]
Just to get some clarification, are you trying to use gossipsub as a request-response protocol or are you using request-response protocol but would like to use it in conjunction with gossipsub? Could you provide any code so we can see what you are trying to do?

If the former, you would have to make sure that either the peers are connected and subscribed to the same topic. If separate topic (thats specific to a peer), you would need to make sure that youre connected to that specific peer to be able to publish to that topic, maintain a separate state per request and respond accordingly. Keep in mind though, while it is possible to do this via gossipsub (I have done this in a couple of projects, including some im working on currently), it is not always best to use it that way but that really falls depends on your use case.

If the latter, this depends on your setup. If your request-response implementation includes gossipsub, you would need to make sure that youre matching the methods to the behaviour (eg `NetworkBehaviour::handle_pending_outbound_connection` should call both behaviour methods, etc) and that youre polling both behaviours. If youre code use both protocols separately then it really depends on your implementation along with making sure the peers are connected and subscribed to said topic so when you send a request or response via request-response, that you can publish accordingly.

If you could provide any code or example of your implementation, that would be great.
